### DISCUSSION TITLE
Raspberry Pi Cluster Docker Swarm

### DISCUSSION THREAD

**Discussion by ats5662** [idaholab/moose#20149] [Q&A General] [answered]
Hey everyone,

I recently started working on a side project trying to figure out how to run MOOSE simulations on a cluster of Raspberry Pis. I got my inspiration for this from [LANL's 750 node Raspberry Pi cluster](https://www.top500.org/news/lanl-turns-to-raspberry-pi-for-supercomputing-solution/). Anyway, I am currently trying to use Docker Swarm to handle the load balancing between the nodes, but I am stuck on not being able to deploy the container as a swarm service. Has anyone done this before or have advice on how to proceed? More details below

**Step 1**
Build docker image of MOOSE specifically for arm v7. Docker hub link below, I have been able to run this on a single raspberry pi without issues. The operating system is Raspberry Pi OS Lite.

[https://hub.docker.com/repository/docker/ats5662/moose_armv7_docker_image](https://hub.docker.com/repository/docker/ats5662/moose_armv7_docker_image)

**Step 2**
Create swarm on 4 node Raspberry Pi cluster  
`sudo docker swarm init --advertise-addr <node 1 IP address> # run on node 1`
`sudo docker swarm join-token manager # add another manager`
`sudo docker swarm join <manager token> # add node 2 as manager`
`sudo docker swarm join <worker token> # add node 3 as worker`
`sudo docker swarm join <worker token> # add node 4 as worker`
`sudo docker node ls # check node list`

**Step 3**
Pull docker image on node 1
`sudo docker pull ats5662/moose_armv7_docker_image`

**Step 4**
Create docker service 
`sudo docker service create --name MOOSE  ats5662/moose_armv7_docker_image`

I am getting the following error/output - this will loop until I stop it
overall progress: 0 out of 1 tasks
1/1: preparing [=================================>                 ]
verify: Detected task failure --> Fail

I am not sure what's wrong because the error message doesn't provide helpful info. If anyone has suggestions on what to do next, it would be greatly appreciated. I am new to docker and docker swarm, so if I am making any rookie mistakes or have a misunderstanding, I apologize. Let me know if additional info is needed. Thanks!

**Comment by GiudGiud** [comment]
@dschwen has experience on MOOSE for raspberry Pi
@milljm for comments on docker & MOOSE
I'd be concerned with using MPI with docker but have no experience on this

**Reply by GiudGiud** [reply]
@loganharbour knows about the docker build a lot too

**Comment by loganharbour** [comment, marked as answer]
This seems more like a docker-not-giving-you-any-useful-information issue. Take a look at https://stackoverflow.com/questions/53222530/getting-detected-task-failure-for-some-images-while-creating-docker-service; looks like you should be able to get more info from the logs?

**Reply by ats5662** [reply]
Thank you for this information. I ended up dropping the docker swarm approach in favor of the [Slurm Workload Manager](https://slurm.schedmd.com/documentation.html) & [openMPI](https://www.open-mpi.org/). I think many real HPCs use similar programs?

These three articles do a pretty good job showing most of the setup process 
Part 1: [https://glmdev.medium.com/building-a-raspberry-pi-cluster-784f0df9afbd](https://glmdev.medium.com/building-a-raspberry-pi-cluster-784f0df9afbd)
Part 2: [https://glmdev.medium.com/building-a-raspberry-pi-cluster-aaa8d1f3d2ca](https://glmdev.medium.com/building-a-raspberry-pi-cluster-aaa8d1f3d2ca)
Part 3: [https://glmdev.medium.com/building-a-raspberry-pi-cluster-f5f2446702e8](https://glmdev.medium.com/building-a-raspberry-pi-cluster-f5f2446702e8)
Job submission: [https://hprc.tamu.edu/files/training/2021/Fall/Slurm_Job_Scheduling_2021_fall.pdf](https://hprc.tamu.edu/files/training/2021/Fall/Slurm_Job_Scheduling_2021_fall.pdf)

I haven't fully tested it out yet, but hopefully will in the coming weeks after school gets out for the summer.
